src/Stream/FileSliceReader.h

```diff
@@ -4,41 +4,165 @@
 #include <string>
 #include <cstddef>
 #include <cstdint>
+#include <stdexcept>
+#include <limits>
 
 namespace Stream
 {
-	// Opens a new file stream that is limited to reading from the provided file slice.
-	class FileSliceReader : public BidirectionalSeekableReader
+	// Creates a stream which is a slice of an existing stream.
+	// Access is bounds checked according to the slice parameters.
+	// The underlying stream must be copy constructible, so that an independent stream can be created.
+	template<class WrappedStreamType>
+	class StreamSlice : public BidirectionalSeekableReader
```

Comment:
What do you think of renaming this to Stream::ReaderSlice? To me, Stream::StreamSlice would indicate a read and write capability (or just be vague on what it does?).

If this sounds good, I can update the filename and the Visual Studio project files.

Reply:
The name probably does need some work. I was purposefully vague on the name since I haven't really decided if this will apply only to read streams, or if it will generalize to write streams as well. I suppose I could just have two templates, though there would be a lot of code in common.

In terms of supporting different levels of the stream interface (Reader, Seekable, BidirectionalSeekable), there may need to be a lot of conditional code added to the template. As such, it may be a simple matter to conditionalize Read/Write, and have the template work for both stream types.

With that said, you're welcome to rename things. We can always rename again later if we come up with a better name.